<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="qgg : Statistical tools for Quantitative Genetic Analyses">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>qgg</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/psoerensen/qgg">View on GitHub</a>

          <h1 id="project_title">qgg</h1>
          <h2 id="project_tagline">Statistical tools for Quantitative Genetic Analyses</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/psoerensen/qgg/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/psoerensen/qgg/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="genomic-feature-linear-mixed-model-analyses" class="anchor" href="#genomic-feature-linear-mixed-model-analyses" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Genomic Feature Linear Mixed Model Analyses</h3>

<p>The qgg package provides a range of genomic feature linear mixed modeling approaches for predicting quantitative trait phenotypes from high resolution genomic polymorphism data. Genomic features are regions on the genome that are hypothesized to be enriched for causal variants affecting the trait. Several genomic feature classes can be formed based on previous studies and different sources of information including genes, chromosomes, biological pathways, gene ontologies, sequence annotation, prior QTL regions, or other types of external evidence. Using prior information on genomic features is important because prediction is difficult for populations of unrelated individuals when the number of causal variants is low relative to the total number of polymorphisms, and causal variants individually have small effects on the traits. The models were implemented using likelihood or Bayesian methods. </p>

<h3>
<a id="genomic-feature-best-linear-unbiased-prediction-models" class="anchor" href="#genomic-feature-best-linear-unbiased-prediction-models" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Genomic feature best linear unbiased prediction models</h3>

<p>Genomic feature best linear unbiased prediction (GFBLUP) models can be fitted. We have demonstrated that the GFBLUP model using prior information on Gene Ontology categories can increase prediction accuracy (compared to existing models) for three quantitative traits in the unrelated, sequenced inbred lines of the <em>Drosophila melanogaster</em> Genetic Reference Panel. These results were supported by simulation studies further illustrating the impact of trait- and genomic feature-specific factors on prediction accuracy. We have extended these models to include multiple features and multiple traits and are currently applying these to a range of quantitative traits in dairy cattle and pigs, and disease traits in human. Different genetic models (e.g. additive, dominance, gene by gene and gene by environment interactions) can be used. Further extensions include a weighted GFBLUP model using differential weighting of the individual genetic marker relationships. </p>

<h3>
<a id="bayesian-multiple-trait-and-multiple-genomic-feature-models" class="anchor" href="#bayesian-multiple-trait-and-multiple-genomic-feature-models" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Bayesian multiple trait and multiple genomic feature models</h3>

<p>Bayesian multiple trait and multiple genomic feature models can be fitted. The models are implemented using an empirical Bayesian method that handles multiple features and multiple traits. The models were implemented using spectral decomposition that plays an important computational role in the Markov chain Monte Carlo strategy. This is a very flexible and formal statistical framework for using prior information to decompose genomic (co)variances and predict trait phenotypes. We applied the model to reveal a strong genetic control of environmental variation of quantitative traits of <em>Drosophila melanogaster</em> using whole-genome sequence data. Furthermore we are currently extending these models to binary outcomes in relation to human diseases and to evaluate different strategies for optimal use of prior information.  </p>

<h3>
<a id="gblup-model-derived-snp-set-tests" class="anchor" href="#gblup-model-derived-snp-set-tests" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>GBLUP model derived SNP set tests</h3>

<p>We have developed GBLUP model derived SNP set tests. The premise of the genomic feature models presented above is that genomic features are enriched for causal variants affecting the traits. However, in reality, the number, location and effect sizes of the true causal variants in the genomic feature are unknown. Therefore we have developed and evaluated a number of SNP set tests derived from a standard Genomic BLUP model. Despite the GBLUP model being considered as a “black box modelling approach” we have shown, using simulations, that it is possible to derive powerful SNP set tests for identifying genomic features enriched for causal variants. These approaches are computationally very fast allowing us to rapidly analyze different layers of genomic feature classes to discover genomic features potentially enriched for causal variants.  Results from these analyses are built into the above mentioned prediction models. We have applied these tests to genotypes and quantitative traits obtained from dairy cattle, pigs, <em>Drosophila melanogaster</em>, and disease traits in human. We are currently extending the Genomic BLUP derived set tests for multiple and longitudinal traits. </p>

<h3>
<a id="references" class="anchor" href="#references" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>References</h3>

<p>Rohde PD, Gaertner B, Ward K, Sørensen P, Mackay TFC. 2017. Genomic analysis of genotype by social environment interaction for <em>Drosophila melanogaster</em> aggressive behavior. Genetics 206:1-16. doi:10.1534/genetics.117.200642.<br>
<a href="http://www.genetics.org/content/early/2017/05/26/genetics.117.200642">http://www.genetics.org/content/early/2017/05/26/genetics.117.200642</a>.</p>        
        
<p>Sørensen IF, Rohde PD, Edwards SM ,Sørensen P. 2017.  Multiple trait covariance association test identifies gene ontology categories associated with chill coma recovery time in <em>Drosophila melanogaster</em>. Scientic Reports 7:2413. doi:10.1038/s41598-017-02281-3.<br>
<a href="https://www.nature.com/articles/s41598-017-02281-3">https://www.nature.com/articles/s41598-017-02281-3</a>. </p>

<p>Fang L, Sahana G, Ma P, Su G, Yu Y, Zhang S, Lund, MS, Sørensen P. 2017. Exploring the genetic architecture and improving genomic prediction accuracy for mastitis and milk production traits in dairy cattle by mapping variants to hepatic transcriptomic regions responsive to intra-mammary infection. Genetics, Selection, Evolution 49(44):1-18. doi:10.1186/s12711-017-0319-0.<br> 
<a href="https://gsejournal.biomedcentral.com/articles/10.1186/s12711-017-0319-0"> https://gsejournal.biomedcentral.com/articles/10.1186/s12711-017-0319-0</a>.</p>        

<p>Fang L, Sahana G, Su G, Yu Y, Zhang S, Lund MS, Sørensen P. 2017. Integrating sequence-based GWAS and RNA-Seq provides novel insights into the genetic basis of mastitis and milk production in dairy cattle. Scientific Reports 7:45560. doi:10.1038/srep45560.<br> 
<a href="https://www.nature.com/articles/srep45560"> https://www.nature.com/articles/srep45560</a>.</p>  
        
<p>Rohde PD, Demontis D, The GEMS Group, Cuyabano BCD, Børglum AD, Sørensen P. 2016. Covariance Association Test (CVAT) identify genetic markers associated with schizophrenia in functionally associated biological processes. Genetics 203:1901–1913. doi:10.1534/genetics.116.189498.<br> 
<a href="http://www.genetics.org/content/203/4/1901">http://www.genetics.org/content/203/4/1901</a>. </p>
        
<p>Edwards SM, Sørensen IF, Sarup P, Mackay TF, Sørensen P. 2016. Genomic prediction for quantitative traits is improved by mapping variants to gene ontology categories in <em>Drosophila melanogaster</em>. Genetics 203:1871-1883. doi:10.1534/genetics.116.187161.<br>
<a href="http://www.genetics.org/content/203/4/1871">http://www.genetics.org/content/203/4/1871</a>. </p>

<p>Sarup P, Jensen J, Ostersen T, Henryon M, Sørensen P. 2016. Increased prediction accuracy using a genomic feature model including prior information on quantitative trait locus regions in purebred Danish Duroc pigs. BMC Genetics 17(1):11. doi:10.1186/s12863-015-0322-9.<br>
 <a href="https://bmcgenet.biomedcentral.com/articles/10.1186/s12863-015-0322-9">https://bmcgenet.biomedcentral.com/articles/10.1186/s12863-015-0322-9</a>. </p>

<p>Ehsani A, Janss L, Pomp D, Sørensen P. 2016. Decomposing genomic variances in a mouse F2 population using information from GWA, GWE and eQTL analyses. Animal Genetics. 47(2):165-173. doi: 10.1111/age.12396.<br>
 <a href="http://onlinelibrary.wiley.com/doi/10.1111/age.12396/abstract"> http://onlinelibrary.wiley.com/doi/10.1111/age.12396/abstract</a>.</p>
                
<p>Sørensen P, de los Campos G, Morgante F, Mackay TFM, Sorensen DA. 2015. Genetic control of environmental variation of two quantitative traits of <em>Drosophila melanogaster</em> revealed by whole-genome sequencing. Genetics 201(2):487-97. doi:10.1534/genetics.115.180273.<br>
  <a href="http://www.genetics.org/content/201/2/487">http://www.genetics.org/content/201/2/487</a>.</p>
       
<p>Edwards SM, Thomsen B, Madsen P, Sørensen P. 2015. Partitioning of genomic variance reveals biological pathways associated with udder health and milk production traits in dairy cattle. Genetics Selection Evolution 47:60. doi:10.1186/s12711-015-0132-6.<br>
 <a href="http://www.gsejournal.org/content/47/1/60"> http://www.gsejournal.org/content/47/1/60</a>.</p>

<p>Skarman A, Shariati M, Janns L, Jiang L, Sørensen P. 2012. A Bayesian variable selection procedure for ranking overlapping gene sets. BMC Bioinformatics 13:73. doi:10.1186/1471-2105-13-73.<br>
  <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-73"> https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-73</a>. </p>

<p>Ehsani A, Sørensen P, Pomp D, Janns L. 2012  Inferring genetic architecture of complex traits using Bayesian integrative analysis of genome and transcriptome data. BMC Genomics 13:456. doi:10.1186/1471-2164-13-456.<br>
  <a href="https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-13-456">https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-13-456</a>.</p>
        
<h3>
<a id="support" class="anchor" href="#support" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Need support?</h3>

 <p>Contact information:<br></p>
  <p>Peter Sørensen: <a href="mailto:pso@mbg.au.dk" target="_top">Send Mail</a> <br></p>    
  <p>Palle Duun Rohde: <a href="mailto:palle.d.rohde@mbg.au.dk" target="_top">Send Mail</a> </p>  

        
    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">qgg maintained by <a href="https://github.com/psoerensen">psoerensen</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
